### PR TITLE
Update arg parsing to clap v3

### PR DESCRIPTION
**Pull request by tranzystorek-io**
Slightly breaking change: trailing arguments are now marked as `last` i.e. they have to be put after a `--` marker, which seems to be more manageable/reasonable.

### CODE REVIEW

File: src/bin/cargo-flamegraph.rs [hunk 1]
```diff
@@ -120,44 +79,44 @@ fn build(opt: &Opt, kind: impl IntoIterator<Item = String>) -> anyhow::Result<Ve
         cmd.arg("--release");
     }
 
-    if let Some(ref package) = opt.package {
+    if let Some(package) = &opt.package {
```
Comment (djc) [inline src/bin/cargo-flamegraph.rs:82]: These changes seem unrelated (that is, unnecessary for porting to clap). Mind putting them in a separate PR or commit?

File: src/bin/flamegraph.rs [hunk 2]
```diff
@@ -1,47 +1,43 @@
...
 struct Opt {
     /// Profile a running process by pid
-    #[structopt(short = "p", long = "pid")]
+    #[clap(short, long)]
     pid: Option<u32>,
 
     /// Generate shell completions for the given shell.
-    #[structopt(long = "completions", conflicts_with = "pid")]
-    completions: Option<structopt::clap::Shell>,
+    #[clap(long, value_name = "SHELL", conflicts_with_all = &["pid", "trailing-arguments"])]
```
Comment (djc) [inline src/bin/flamegraph.rs:17]: Why spell out the `value_name` here?

File: src/bin/flamegraph.rs [hunk 3]
```diff
@@ -1,47 +1,43 @@
...
+    completions: Option<Shell>,
 
-    #[structopt(flatten)]
+    #[clap(flatten)]
     graph: flamegraph::Options,
 
-    #[structopt(parse(from_os_str), long = "perfdata", conflicts_with = "pid")]
+    #[clap(long = "perfdata", conflicts_with = "pid", parse(from_os_str))]
```
Comment (djc) [inline src/bin/flamegraph.rs:23]: Is reordering necessary here or just a style tweak?

**Review by djc — COMMENTED** [review]
Thanks for working on this! In order to make this easy to review, I'd like to have just the minimal changes required to get it compiling with clap, and then have a follow-up PR (or follow-up commits in this PR) to further improve the command-line parsing.

Also could you please elaborate with some examples on things that are now allowed or are no longer allowed with regards to the argument parsing?

File: src/bin/flamegraph.rs (hunk 2 above)
Comment (tranzystorek-io) [reply src/bin/flamegraph.rs:17]: It describes what we're actually specifying, i.e. the target shell, otherwise clap just copies the field name

File: src/bin/flamegraph.rs (hunk 3 above)
Comment (tranzystorek-io) [reply src/bin/flamegraph.rs:23]: The latter

**Comment by tranzystorek-io** [comment]
Thank you for having me! 😄

As for the new things, clap 3 mostly just cleans up all the APIs and incorporates old structopt under the `derive` umbrella.

Among the more interesting new additions is the `ArgEnum` trait which simplifies using enums as arguments, e.g.:

```rust
use clap::ArgEnum;

#[derive(ArgEnum)]
enum ProcessingMode {
    Batch,
    Interactive,
}

#[derive(Parser)]
struct Cli {
    #[long, arg_enum]
    mode: ProcessingMode,
}
```

This would accept args like `myprogram --mode batch` without any manual impls like `FromStr`, as well as display possible variants in the help message. The variant names can also be switched similarly to serde, to `lowercase`, `snake_case`, `kebab-case` etc.

**Comment by tranzystorek-io** [comment]
I went ahead and disabled the CLI portion of `inferno` since it introduced unused clap 2 dependency

**Comment by tranzystorek-io** [comment]
> Also could you please elaborate with some examples on things that are now allowed or are no longer allowed with regards to the argument parsing?

I seem to have missed the point of your question, I'm assuming you were asking about the breaking changes I introduced.

Now that the `trailing_arguments` in both `flamegraph` and `cargo_flamegraph` are marked as `last`, they **must** be placed after the `--`, e.g.:

```console
cargo flamegraph -F 10000 -- --my --args -a -b -c
```

This is slightly more restrictive, but also now the user unambiguously knows how to pass any arguments to their own application.

**Comment by djc** [comment]
Thanks!
